### PR TITLE
Once a Connection is Established, Respond with 'Connection Established'

### PR DESCRIPTION
**Pull request by DeKaylaPhillips**
# Once a Connection is Established, Respond with 'Connection Established'

## Related Tickets & Documents

- Ticket: https://trello.com/c/Lh63kMhy/42-once-a-connection-is-established-respond-with-connection-established

Screen Shot 2023-07-18 at 3 10 00 PM

## Type of PR 

- [X] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This branch refactors the TCP server to allow a client connected at localhost:3000 to receive raw data on the socket from the server after the connection has been established. The project configuration file in this branch has also been updated to resolve issues with the Jest testing framework.

Screen Shot 2023-07-18 at 3 21 41 PM

## Added/updated tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

### CODE REVIEW

File: tests/server/server.spec.ts [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
   type MaybeNumber = number | undefined
   type MaybeString = string | undefined
   type MaybeVoidFunction = (() => void) | undefined
+  type SocketCallback = ((socket: any) => void) | undefined
+
+  class MockSocket {
+    write: (
+      data: string | Uint8Array
+    ) => boolean = jest.fn()
+  }
```
Comment (Naomi-Dennis) [inline tests/server/server.spec.ts:15]: 🎉

File: tests/server/server.spec.ts [hunk 2]
```diff
@@ -31,15 +59,33 @@ describe('Server', () => {
...
+    server.openTCPConnection()
+    expect(logMock).toHaveBeenCalledTimes(2)
+    expect(logMock).toHaveBeenCalledWith('Client has connected.')
+  })
+
+  test('responds to a client connected to the TCP server', () => {
+    hasMockClientConnected = true
+    server.openTCPConnection()
+    expect(server.socket.write).toHaveBeenCalledTimes(1)
+    expect(server.socket.write).toHaveBeenCalledWith('Connection established.')
```
Comment (Naomi-Dennis) [inline tests/server/server.spec.ts]: 🎉 

nitpick: Instead of using `toHaveBeenCalledTimes(1)` you can use `toHaveBeenCalled()`

**Comment by Naomi-Dennis** [comment]
Nice job! I know this one was a bit rough, but you tested the `socket.write` part wonderfully :)

**Review by Naomi-Dennis — APPROVED** [review]
🚢 🚢 🚢

File: tests/server/server.spec.ts (hunk 2 above)
Comment (DeKaylaPhillips) [reply tests/server/server.spec.ts]: Thank you! I went ahead and updated the matcher! :)

**Comment by DeKaylaPhillips** [comment]
> Nice job! I know this one was a bit rough, but you tested the `socket.write` part wonderfully :)

Thank you a bunch, Naomi!

**Pull request merged**